propolis-server/src/api.rs

```diff
@@ -0,0 +1,189 @@
+}
+
+/// Requested state of an Instance.
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub struct InstanceStateChange {
+    pub state: InstanceStateRequested,
+}
+
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub enum InstanceStateRequested {
+    Run,
+    Stop,
+    Reboot,
+}
+
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub struct InstanceSerialRequest {
+    pub bytes: Vec<u8>,
+}
+
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub struct InstanceSerialResponse {
+    pub bytes: Vec<u8>,
+}
+
+/// Current state of an Instance.
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub enum InstanceState {
+    Creating,
+    Starting,
+    Running,
+    Stopping,
+    Stopped,
+    Repairing,
+    Failed,
+    Destroyed,
+}
+
+#[derive(Clone, Deserialize, Serialize, JsonSchema)]
+pub struct InstanceProperties {
+    pub generation_id: u64,
+
+    /// Unique identifier for this Instance.
+    pub id: Uuid,
+    /// Human-readable name of the Instance.
+    pub name: String,
+    /// Free-form text description of an Instance.
+    pub description: String,
+
+    /// ID of the image used to initialize this Instance.
+    pub image_id: Uuid,
+    /// ID of the bootrom used to initialize this Instance.
+    pub bootrom_id: Uuid,
+    /// Size of memory allocated to the Instance, in MB.
+    pub memory: u64,
+    /// Number of vCPUs to be allocated to the Instance.
+    pub vcpus: u8,
+}
+
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub struct Instance {
+    pub properties: InstanceProperties,
+    pub state: InstanceState,
+
+    pub disks: Vec<DiskAttachment>,
+    pub nics: Vec<NicAttachment>,
+}
+
+/// Describes how to connect to one or more storage agent services.
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub struct StorageAgentDescription {
+    /// Addresses of storage agents.
+    pub agents: Vec<std::net::SocketAddrV6>,
+
+    /// Opaque key material for encryption and decryption.
+    /// May become more structured as encryption scheme is solidified.
+    pub key: Vec<u8>,
+
+    /// Minimum number of redundant copies of a block which must
+    /// be written until data is considered "persistent".
+    pub write_redundancy_threshold: u32,
+}
+
+/// Refer to RFD 135 for more information on Virtual Storage Interfaces.
+/// This describes the type of disk which should be exposed to the guest VM.
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub enum DiskType {
+    NVMe,
+    VirtioBlock,
+}
+
+/// Describes a virtual disk.
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub struct Disk {
+    /// Unique identifier for this disk.
+    pub id: Uuid,
+
+    /// Storage agents which implement networked block device servers.
+    pub storage_agents: StorageAgentDescription,
+
+    /// Size of the disk (blocks).
+    pub block_count: u64,
+
+    /// Block size (bytes).
+    pub block_size: u32,
+
+    /// Storage interface.
+    pub interface: DiskType,
+}
+
+// TODO: Struggling to make this struct derive "JsonSchema"
```

Comment:
It would be nice for the disk flags to be... well, flags, since there may be any combination of them. Agreed that we *could* use a u32 for this and then do the conversion ourselves (which is what we're more-or-less currently doing) to a stronger type.

However, that leaves the (minor) problem of "the u32 could be interpreted as something else by the server", or "something that was not `DiskFlags` could be supplied by the caller".

Totally agreed that "just use u32 on the wire" is a valid workaround, but avoiding the type conversion (if bitflags could derive `JsonSchema`) would remove the possibility of bad type casting.

Reply:
I agree with all that, for sure. Maybe I put this comment in an unhelpful place, but I was trying to ask why the workaround below specified the flag as `const u32` rather than just defining `enum DiskFlags` with the appropriate variants and `From` impl. But I think I understand now -- you're writing it in a way such that, if the `bitflags!` bug can be resolved, the consuming code need not change. Good idea, apologies for the confusing comment here!